flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java

```diff
@@ -818,71 +800,46 @@ else if (numConsumers == 0) {
-				}
 			}
 			// ----------------------------------------------------------------
 			// Consumer is running => send update message now
+			// Consumer is deploying => cache the partition info which would be
+			// sent after switching to running
 			// ----------------------------------------------------------------
-			else {
-				if (consumerState == RUNNING) {
-					final LogicalSlot consumerSlot = consumer.getAssignedResource();
-
-					if (consumerSlot == null) {
-						// The consumer has been reset concurrently
-						continue;
-					}
-
-					final TaskManagerLocation partitionTaskManagerLocation = partition.getProducer()
-							.getCurrentAssignedResource().getTaskManagerLocation();
-					final ResourceID partitionTaskManager = partitionTaskManagerLocation.getResourceID();
-
-					final ResourceID consumerTaskManager = consumerSlot.getTaskManagerLocation().getResourceID();
-
-					final ResultPartitionID partitionId = new ResultPartitionID(partition.getPartitionId(), attemptId);
-
-					final ResultPartitionLocation partitionLocation;
+			else if (consumerState == DEPLOYING || consumerState == RUNNING) {
+				final LogicalSlot consumerSlot = consumer.getAssignedResource();
+				if (consumerSlot == null) {
+					// The consumer has been reset concurrently
+					continue;
+				}
 
-					if (consumerTaskManager.equals(partitionTaskManager)) {
-						// Consuming task is deployed to the same instance as the partition => local
-						partitionLocation = ResultPartitionLocation.createLocal();
-					}
-					else {
-						// Different instances => remote
-						final ConnectionID connectionId = new ConnectionID(
-								partitionTaskManagerLocation,
-								partition.getIntermediateResult().getConnectionIndex());
+				final TaskManagerLocation partitionTaskManagerLocation = partition.getProducer()
+					.getCurrentAssignedResource().getTaskManagerLocation();
+				final ResourceID partitionTaskManager = partitionTaskManagerLocation.getResourceID();
+				final ResourceID consumerTaskManager = consumerSlot.getTaskManagerLocation().getResourceID();
 
-						partitionLocation = ResultPartitionLocation.createRemote(connectionId);
-					}
+				final ResultPartitionLocation partitionLocation;
+				if (consumerTaskManager.equals(partitionTaskManager)) {
+					// Consuming task is deployed to the same instance as the partition => local
+					partitionLocation = ResultPartitionLocation.createLocal();
+				} else {
+					// Different instances => remote
+					final ConnectionID connectionId = new ConnectionID(
+						partitionTaskManagerLocation,
+						partition.getIntermediateResult().getConnectionIndex());
+					partitionLocation = ResultPartitionLocation.createRemote(connectionId);
+				}
 
-					final InputChannelDeploymentDescriptor descriptor = new InputChannelDeploymentDescriptor(
-							partitionId, partitionLocation);
+				final ResultPartitionID partitionId = new ResultPartitionID(partition.getPartitionId(), attemptId);
```

Comment:
Not sure, how it worked before, but this looks like a bug. All other code around does not make any assumption that this execution is a producer for current partition but for `ResultPartitionID` we use `attemptId` of this execution. Although, we should use `partition.getProducer().getCurrentExecutionAttempt().attemptId`. Looking how `scheduleOrUpdateConsumers` is called in `markFinished`, it also updates consumers of other producers.

Reply:
I checked the process and it seemed not a bug before.

If one execution is marked finished, then this execution should be regarded as a producer to update all its consumers, so it is safe to use `attemptId` before.

Nevertheless, after we migrate the creation of `PartitionInfo` from `Execution` as you suggested, we could use `partition.getProducer().getCurrentExecutionAttempt().getAttemptId()` instead to generate `ResultPartitionID` for simplifying the `PartitionInfo#fromEdge(ExecutionEdge)`.

Reply:
True, it is not a bug, but it was a bit confusing. It seems that we do not use `ExecutionAttemptID` eventually on Task side when the input channel gets updated because the originally created unknown channel already contains the correct one from TaskDD. I think we should reduce `PartitionInfo` to contain only relevant fields which are actually used for the update when we work on `ShuffleService. updatePartitionInfos`.